Scripts/Helpers/ProxyUI.cs

```diff
@@ -0,0 +1,549 @@
+                materialAndAssociatedColors.fourthElement = Color.white * animateFromAlpha; // Encode the alpha for the FROM color value, color.a
+                materialAndAssociatedColors.fifthElement = Color.white * animateToAlpha; // // Encode the alpha for the TO color value, color.a
+            }
+
+            while (currentAmount < kTargetAmount)
+            {
+                var smoothedAmount = MathUtilsExt.SmoothInOutLerpFloat(currentAmount += Time.unscaledDeltaTime * speedScalar);
+                foreach (var materialAndAssociatedColors in materialsAndColors)
+                {
+                    var currentAlpha = Color.Lerp(materialAndAssociatedColors.fourthElement, materialAndAssociatedColors.fifthElement, smoothedAmount);
+                    materialAndAssociatedColors.firstElement.SetFloat(shaderAlphaPropety, currentAlpha.a); // Alpha is encoded in color.a
+                }
+
+                yield return null;
+            }
+        }
+
+        IEnumerator AnimateBodyColorVisibility(bool isVisible)
+        {
+            var bodyVisibilityDefinition = m_AffordanceMap.bodyVisibilityDefinition;
+            foreach (var kvp in m_BodyMaterialOriginalColorMap)
+            {
+                kvp.Value.animateFromValue = kvp.Key.GetColor(bodyVisibilityDefinition.colorProperty);
+            }
+
+            const float kTargetAmount = 1f;
+            const float kHiddenValue = 0.25f;
+            var speedScalar = isVisible ? k_FadeInSpeedScalar : k_FadeOutSpeedScalar;
+            var currentAmount = 0f;
+            var shaderColorPropety = bodyVisibilityDefinition.colorProperty;
+            while (currentAmount < kTargetAmount)
+            {
+                var smoothedAmount = MathUtilsExt.SmoothInOutLerpFloat(currentAmount += Time.unscaledDeltaTime * speedScalar);
+                foreach (var kvp in m_BodyMaterialOriginalColorMap)
+                {
+                    // Set original cached color when visible, transparent when hidden
+                    var valueFrom = kvp.Value.animateFromValue;
+                    var valueTo = isVisible ? kvp.Value.originalValue : new Color(valueFrom.r, valueFrom.g, valueFrom.b, kHiddenValue);
+                    var currentColor = Color.Lerp(valueFrom, valueTo, smoothedAmount);
+                    kvp.Key.SetColor(shaderColorPropety, currentColor);
+                }
+
+                yield return null;
+            }
+
+            // Mandate target values have been set
+            foreach (var kvp in m_BodyMaterialOriginalColorMap)
+            {
+                var originalColor = kvp.Value.originalValue;
+                kvp.Key.SetColor(shaderColorPropety, isVisible ? originalColor : new Color(originalColor.r, originalColor.g, originalColor.b, kHiddenValue));
+            }
+        }
+
+        IEnumerator AnimateBodyAlphaVisibility(bool isVisible)
+        {
+            var bodyVisibilityDefinition = m_AffordanceMap.bodyVisibilityDefinition;
```

Comment:
A lot of these loops are extremely visible and even rely on the same kind of templated data -  any way we can combine these functions?